Editorial review: Document ViewTransition.waitUntil() (#43962)
mdn/content: files/en-us/web/api/viewtransition/waituntil/index.md

---
title: "ViewTransition: waitUntil() method"
short-title: waitUntil()
slug: Web/API/ViewTransition/waitUntil
page-type: web-api-instance-method
browser-compat: api.ViewTransition.waitUntil
---

{{APIRef("View Transition API")}}

The **`waitUntil()`** method of the {{domxref("ViewTransition")}} interface delays finishing the view transition and the destruction of the associated pseudo-element tree until a {{jsxref("Promise")}} passed into the method has resolved.

## Syntax

```js-nolint
waitUntil(promise)
```

### Parameters

- `promise`
  - : A {{jsxref("Promise")}} that, when resolved, causes the view transition to finish and the associated pseudo-element tree to be destroyed. This can be any promise.

### Return value

None ({{jsxref("undefined")}}).

## Description

> When an SPA view transition is started via {{domxref("Document.startViewTransition()")}}, the browser automatically constructs a [pseudo-element tree](/en-US/docs/Web/API/View_Transition_API/Using#the_view_transition_pseudo-element_tree) to display and animate outgoing and inbound changes to the DOM. This tree is constructed when the view transition starts animating and destroyed when the animations associated with all view transition pseudo-elements are in the finished state ({{domxref("ViewTransition.finished", "finished")}} is resolved).

Reviewer: ```suggestion
When a same-document view transition is started (typically via {{domxref("Document.startViewTransition()")}}), the browser automatically constructs a [pseudo-element tree](/en-US/docs/Web/API/View_Transition_API/Using#the_view_transition_pseudo-element_tree) to display and animate outgoing and inbound changes to the DOM. This tree is constructed when the view transition starts animating and is destroyed when the animations associated with all view transition pseudo-elements reach the finished state ({{domxref("ViewTransition.finished", "finished")}} is resolved).
```

we also recently shipped element scoped view transitions (which would not be Document.startViewTransittion()), not sure if it's worthwhile to change that.

Author: I understand your concern. To allow for this, I have tweaked your suggestion above to:

> When a same-document view transition is started (typically via {{domxref("Document.startViewTransition()")}}), the browser...

Before committing it.

> This works fine in most cases, however, some advanced use cases benefit from the pseudo-tree persisting beyond the animation finish state. This can be achieved using the `waitUntil()` method, which is passed a {{jsxref("Promise")}} as an argument. Calling `waitUntil()` causes the pseudo-tree to persist until the promise is resolved. At this point the `finished` promise also resolves.

Reviewer: ```suggestion
This works fine in most cases. However, some advanced use cases benefit from the pseudo-tree persisting beyond the animation finish state. This can be achieved using the `waitUntil()` method, which is passed a {{jsxref("Promise")}} as an argument. Calling `waitUntil()` causes the pseudo-tree to persist until the promise is resolved. At this point, the `finished` promise also resolves.

Repeated calls to `waitUntil()` specifying multiple different promises will delay the finish state until all the given promises are resolved.
```

---

{{APIRef("View Transition API")}}

The **`waitUntil()`** method of the {{domxref("ViewTransition")}} interface delays finishing the view transition and the destruction of the associated pseudo-element tree until a {{jsxref("Promise")}} passed into the method has resolved.

## Syntax

```js-nolint
waitUntil(promise)
```

### Parameters

- `promise`
  - : A {{jsxref("Promise")}} that, when resolved, causes the view transition to finish and the associated pseudo-element tree to be destroyed. This can be any promise.

### Return value

None ({{jsxref("undefined")}}).

## Description

When a same-document view transition is started (typically via {{domxref("Document.startViewTransition()")}}), the browser automatically constructs a [pseudo-element tree](/en-US/docs/Web/API/View_Transition_API/Using#the_view_transition_pseudo-element_tree) to display and animate outgoing and inbound changes to the DOM. This tree is constructed when the view transition starts animating and is destroyed when the animations associated with all view transition pseudo-elements reach the finished state ({{domxref("ViewTransition.finished", "finished")}} is resolved).

> This works fine in most cases. However, some advanced use cases benefit from the pseudo-tree persisting beyond the animation finish state. This can be achieved using the `waitUntil()` method, which is passed a {{jsxref("Promise")}} as an argument. Calling `waitUntil()` causes the pseudo-tree to persist until the promise is resolved. At this point, the `finished` promise also resolves.

Reviewer 2: I'm seeing a lot of the chromestatus here. I'm not entirely sure what makes these count as "advanced" either.

```suggestion
This works well for most use-cases, but there are some that benefit from the pseudo-tree persisting beyond the animation finish state. This can be achieved using the `waitUntil()` method, which is passed a {{jsxref("Promise")}} as an argument. Calling `waitUntil()` causes the pseudo-tree to persist until the promise is resolved. At this point, the `finished` promise also resolves.
```

Repeated calls to `waitUntil()` specifying multiple different promises will delay the finish state until all the given promises are resolved.

Use cases include:

- Combining a view transition with [scroll-driven animations](/en-US/docs/Web/CSS/Guides/Scroll-driven_animations). When the transition animation is controlled by a scroll progress or view progress timeline, the subtree should persist when the animations finish since scrolling back should be able to animate the pseudo-elements in reverse.

> - Combining a view transition with {{domxref("Window.requestAnimationFrame", "requestAnimationFrame()")}}. When you're updating the state of your elements in `requestAnimationFrame()` callbacks, the view transition system does not know how long to persist the pseudo-element tree for, and will do it immediately as soon as the CSS animations are finished.

Reviewer 2: What is "it" in "and will do it immediately "

Author: Ah, "do" was supposed to be "destroy", and "it" refers to the pseudo-element tree. Makes more sense now.

- Any situation in which you want to delay the view transition finishing until an event has occurred. You might for example want to start the view transition on `pointerdown` and not finish it until `pointerup`.

## Examples

> ### Basic `waitUntil()` usage

Reviewer 2: Up to you, but this looks better.

```suggestion
### Basic usage
```

Reviewer 2: Nice example!
